File: internal/plugins/react/rules/jsx_first_prop_new_line/jsx_first_prop_new_line_test.go
package jsx_first_prop_new_line

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/react/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestJsxFirstPropNewLineRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &JsxFirstPropNewLineRule, []rule_tester.ValidTestCase{
		{
			Code: `var x = <div foo="bar" />`,
			Tsx:  true,
		},
		{
			Code: `var x = <div foo="bar" baz="qux" />`,
			Tsx:  true,
		},
		{
			Code: `var x = <div
  foo="bar"
/>`,
			Tsx:     true,
			Options: "always",
		},
		{
			Code:    `var x = <div foo="bar" />`,
			Tsx:     true,
			Options: "never",
		},
		{
			Code: `var x = <div
  foo="bar"
  baz="qux"
/>`,
			Tsx: true,
		},
		{
			// "never": no props is valid
			Code:    `var x = <Foo />`,
			Tsx:     true,
			Options: "never",
		},
		{
			// "multiline-multiprop": single prop on same line is valid
			Code:    `var x = <Foo bar />`,
			Tsx:     true,
			Options: "multiline-multiprop",
		},
		{
			// "multiline-multiprop": multiple props on same single line is valid
			Code:    `var x = <Foo bar baz />`,
			Tsx:     true,
			Options: "multiline-multiprop",
		},
		{
			// "multiline-multiprop": multiline with first prop on new line is valid
			Code: `var x = <Foo
  foo={{}}
  bar
/>`,
			Tsx:     true,
			Options: "multiline-multiprop",
		},
		{
			// "multiprop": single prop is valid on same line
			Code:    `var x = <Foo bar />`,
			Tsx:     true,
			Options: "multiprop",
		},
		{
			// "multiprop": no props is valid
			Code:    `var x = <Foo />`,
			Tsx:     true,
			Options: "multiprop",
		},
		{
			// "always": no props is valid
			Code:    `var x = <Foo />`,
			Tsx:     true,
			Options: "always",
		},
		{
			// "multiline": single-line single prop is valid
			Code:    `var x = <Foo bar />`,
			Tsx:     true,
			Options: "multiline",
		},
		{
			// "multiline": multiline with first prop on new line is valid
			Code: `var x = <Foo
  propOne="one"
  propTwo="two"
/>`,
			Tsx:     true,
			Options: "multiline",
		},
	}, []rule_tester.InvalidTestCase{
		{
			// Default mode (multiline-multiprop): multiline with multiple props, first on same line
			Code: `var x = <Foo bar baz
/>`,
			Tsx: true,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{`var x = <Foo
bar baz
/>`},
		},
		{
			Code:    `var x = <div foo="bar" />`,
			Tsx:     true,
			Options: "always",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{"var x = <div\nfoo=\"bar\" />"},
		},
		{
			Code: `var x = <div
  foo="bar"
/>`,
			Tsx:     true,
			Options: "never",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnSameLine",
					Line:      2,
					Column:    3,
				},
			},
			Output: []string{`var x = <div foo="bar"
/>`},
		},
		{
			Code: `var x = <div foo="bar"
  baz="qux"
/>`,
			Tsx:     true,
			Options: "multiprop",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{`var x = <div
foo="bar"
  baz="qux"
/>`},
		},
		{
			// "always": multiline first prop on same line
			Code:    `var x = <Foo propOne="one" propTwo="two" />`,
			Tsx:     true,
			Options: "always",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{"var x = <Foo\npropOne=\"one\" propTwo=\"two\" />"},
		},
		{
			// "multiline": prop on same line when multiline
			Code: `var x = <Foo prop={{}}
/>`,
			Tsx:     true,
			Options: "multiline",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{`var x = <Foo
prop={{}}
/>`},
		},
		{
			// "multiline-multiprop": multiline with first prop on same line
			Code: `var x = <Foo bar={{}} baz
/>`,
			Tsx:     true,
			Options: "multiline-multiprop",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{`var x = <Foo
bar={{}} baz
/>`},
		},
		{
			// "multiprop": multiple props on same line should error
			Code:    `var x = <Foo propOne="one" propTwo="two" />`,
			Tsx:     true,
			Options: "multiprop",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnNewLine",
					Line:      1,
					Column:    14,
				},
			},
			Output: []string{"var x = <Foo\npropOne=\"one\" propTwo=\"two\" />"},
		},
		{
			// "multiprop": single prop on new line in multiline element should be on same line
			Code: `var x = <Foo
  bar
/>`,
			Tsx:     true,
			Options: "multiprop",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "propOnSameLine",
					Line:      2,
					Column:    3,
				},
			},
			Output: []string{`var x = <Foo bar
/>`},
		},
	})
}

File: packages/rslint-test-tools/tests/eslint-plugin-stylistic/rules/jsx-first-prop-new-line.test.ts
/**
 * @fileoverview Ensure proper position of the first property in JSX.
 * @author Joachim Seminck
 *
 * Ported verbatim from @stylistic/eslint-plugin v5.10.0:
 *   packages/eslint-plugin/rules/jsx-first-prop-new-line/jsx-first-prop-new-line.test.ts
 *
 * Transformations applied per the porting spec:
 *  - `run({ name, rule, parserOptions, ... })` ->
 *    `ruleTester.run('jsx-first-prop-new-line', null as never, { valid, invalid })`
 *  - `parserOptions` (`ecmaFeatures.jsx: true`) dropped — rslint resolves via
 *    tsconfig and the RuleTester routes JSX fixtures to `.tsx`.
 *  - Per-case `features` arrays dropped (see note below).
 *
 * The upstream file wraps its cases in the `valids()` / `invalids()` helpers from
 * `shared/test-utils/parsers-jsx.ts`. Those helpers multiplex each case across
 * several PARSERS (ESLint-default, @babel/eslint-parser, @typescript-eslint/parser)
 * and append a `// features: [...], parser: ...` comment to `code`/`output`. With
 * the resolved toolchain (ESLint 10.5.0) the babel variant is always skipped
 * (`skipBabel = gte(ESLint.version, '10.0.0')` === true). The lone `features:
 * ['ts', 'no-babel-old']` case (`<DataTable<Items> .../>`) additionally skips the
 * default-parser variant (`ts` ∈ skipBase) and runs ONLY on @typescript-eslint —
 * which is exactly the parser rslint's ts-go emulates for a `.tsx` fixture. Every
 * other case runs on default + @typescript-eslint, identical under ts-go. So each
 * case is ported ONCE as its literal source (the appended parser-comment is
 * dropped); the JSX code, including the multi-line plain-backtick templates with
 * their leading newline + indentation, is preserved verbatim and runs as `.tsx`.
 *
 * The expected messages resolve from the plugin's own `meta.messages`:
 *   propOnNewLine:  "Property should be placed on a new line"
 *   propOnSameLine: "Property should be placed on the same line as the component declaration"
 *
 * The upstream file contains NO `$` unindent template tags (the multi-line cases
 * use plain backtick literals — indentation preserved verbatim), NO `readFileSync`
 * external-fixture cases, NO `suggestions`, and only the single `run()` block
 * above. The `._css_` / `._json_` / `._markdown_` test files don't exist for this
 * rule.
 *
 * No case surfaces a rslint<->upstream gap, so nothing is moved to KNOWN GAPS.
 */

import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('jsx-first-prop-new-line', null as never, {
  valid: [
    {
      code: '<Foo />',
      options: ['never'],
    },
    {
      code: '<Foo prop="bar" />',
      options: ['never'],
    },
    {
      code: '<Foo {...this.props} />',
      options: ['never'],
    },
    {
      code: '<Foo a a a />',
      options: ['never'],
    },
    {
      code: `
        <Foo a
          b
        />
      `,
      options: ['never'],
    },
    {
      code: '<Foo />',
      options: ['multiline'],
    },
    {
      code: '<Foo prop="one" />',
      options: ['multiline'],
    },
    {
      code: '<Foo {...this.props} />',
      options: ['multiline'],
    },
    {
      code: '<Foo a a a />',
      options: ['multiline'],
    },
    {
      code: `
        <Foo
          propOne="one"
          propTwo="two"
        />
      `,
      options: ['multiline'],
    },
    {
      code: `
        <Foo
          {...this.props}
          propTwo="two"
        />
      `,
      options: ['multiline'],
    },
    {
      code: `
        <Foo bar />
      `,
      options: ['multiline-multiprop'],
    },
    {
      code: `
        <Foo bar baz />
      `,
      options: ['multiline-multiprop'],
    },
    {
      code: `
        <Foo prop={{
        }} />
      `,
      options: ['multiline-multiprop'],
    },
    {
      code: `
        <Foo
          foo={{
          }}
          bar
        />
      `,
      options: ['multiline-multiprop'],
    },
    {
      code: '<Foo />',
      options: ['always'],
    },
    {
      code: `
        <Foo
          propOne="one"
          propTwo="two"
        />
      `,
      options: ['always'],
    },
    {
      code: `
        <Foo
          {...this.props}
          propTwo="two"
        />
      `,
      options: ['always'],
    },
    {
      code: `
        <Foo />
      `,
      options: ['multiprop'],
    },
    {
      code: `
        <Foo bar />
      `,
      options: ['multiprop'],
    },
    {
      code: `
        <Foo {...this.props} />
      `,
      options: ['multiprop'],
    },
  ],

  invalid: [
    {
      code: `
        <Foo propOne="one" propTwo="two" />
      `,
      output: `
        <Foo
propOne="one" propTwo="two" />
      `,
      options: ['always'],
      errors: [{ messageId: 'propOnNewLine' }],
    },
    {
      code: `
        <Foo propOne="one"
          propTwo="two"
        />
      `,
      output: `
        <Foo
propOne="one"
          propTwo="two"
        />
      `,
      options: ['always'],
      errors: [{ messageId: 'propOnNewLine' }],
    },
    {
      code: `
        <Foo
          propOne="one"
          propTwo="two"
        />
      `,
      output: `
        <Foo propOne="one"
          propTwo="two"
        />
      `,
      options: ['never'],
      errors: [{ messageId: 'propOnSameLine' }],
    },
    {
      code: `
        <Foo prop={{
        }} />
      `,
      output: `
        <Foo
prop={{
        }} />
      `,
      options: ['multiline'],
      errors: [{ messageId: 'propOnNewLine' }],
    },
    {
      code: `
        <Foo bar={{
        }} baz />
      `,
      output: `
        <Foo
bar={{
        }} baz />
      `,
      options: ['multiline-multiprop'],
      errors: [{ messageId: 'propOnNewLine' }],
    },
    {
      code: `
      <Foo propOne="one" propTwo="two" />
      `,
      output: `
      <Foo
propOne="one" propTwo="two" />
      `,
      options: ['multiprop'],
      errors: [{ messageId: 'propOnNewLine' }],
    },
    {
      code: `
      <Foo
bar />
      `,
      output: `
      <Foo bar />
      `,
      options: ['multiprop'],
      errors: [{ messageId: 'propOnSameLine' }],
    },
    {
      code: `
      <Foo
{...this.props} />
      `,
      output: `
      <Foo {...this.props} />
      `,
      options: ['multiprop'],
      errors: [{ messageId: 'propOnSameLine' }],
    },
    {
      code: `
        <DataTable<Items> fullscreen keyField="id" items={items}
          activeSortableColumn={sorting}
          onSortClick={handleSortedClick}
          rowActions={[
          ]}
        />
      `,
      output: `
        <DataTable<Items>
fullscreen keyField="id" items={items}
          activeSortableColumn={sorting}
          onSortClick={handleSortedClick}
          rowActions={[
          ]}
        />
      `,
      options: ['multiline'],
      errors: [{ messageId: 'propOnNewLine' }],
    },
  ],
});
